File: BankImport/BankImport/Slf.Dms.Service.BankImport.Slf.Dms.Service.BankImport/ImportMain.vb
Imports Drg.Util.DataAccess
Imports Drg.Util.DataHelpers
Imports EnterpriseDT.Net.Ftp
Imports Microsoft.VisualBasic.CompilerServices
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Configuration
Imports System.Data
Imports System.Data.SqlClient
Imports System.Diagnostics
Imports System.IO
Imports System.Net.Mail
Imports System.Reflection
Imports System.Security.Cryptography
Imports System.Windows.Forms

Namespace Slf.Dms.Service.BankImport.Slf.Dms.Service.BankImport
    Public Class ImportMain
        ' Methods
        Private Shared Sub EmailLog(ByVal log As String)
            If (IIf(((Not ImportMain.LogEmailTo Is Nothing) AndAlso (ImportMain.LogEmailTo.Length > 0)), 1, 0) <> 0) Then
                Dim message As New MailMessage("service@dmsisupport.com", ImportMain.LogEmailTo)
                Dim client As New SmtpClient(ImportMain.SmtpServer)
                message.Subject = ("Import Log for " & DateTime.Now.ToShortDateString)
                message.Body = log
                Try 
                    client.Send(message)
                Catch exception1 As Exception
                    ProjectData.SetProjectError(exception1)
                    Dim exception As Exception = exception1
                    Trace.WriteLine("Exception sending mail:")
                    Trace.WriteLine(exception.ToString)
                    ProjectData.ClearProjectError
                End Try
            End If
        End Sub

        Private Shared Function GenerateMD5(ByVal fileName As String) As Byte()
            Using md As MD5 = MD5.Create
                Return md.ComputeHash(File.ReadAllBytes(fileName))
            End Using
        End Function

        Private Shared Function GetFtpLatest() As String()
            Trace.WriteLine("--------------------------------")
            Trace.Write(DateTime.Now.ToString)
            Trace.Write(" - Retrieving from FTP" & ChrW(13) & ChrW(10))
            Trace.Write("Server: ")
            Trace.Write(ImportMain.FtpServer)
            Trace.Write(ChrW(13) & ChrW(10) & "Folder: ")
            Trace.Write(ImportMain.FtpFolder)
            Trace.Write(ChrW(13) & ChrW(10) & "--------------------------------" & ChrW(13) & ChrW(10))
            Try
                Trace.WriteLine("Connecting...")
                Dim client As New FTPClient(ImportMain.FtpServer)
                Trace.WriteLine("Logging in...")
                client.Login(ImportMain.FtpUserName, ImportMain.FtpPassword)
                Trace.WriteLine("Setting up passive, ASCII transfers...")
                client.ConnectMode = FTPConnectMode.PASV
                client.TransferType = FTPTransferType.ASCII
                If (IIf(((Not ImportMain.FtpFolder Is Nothing) AndAlso (ImportMain.FtpFolder.Length > 0)), 1, 0) <> 0) Then
                    client.ChDir(ImportMain.FtpFolder)
                End If
                Dim textArray2 As String() = client.Dir
                Dim list As New ArrayList(textArray2.Length)
                Dim num2 As Integer = (textArray2.Length - 1)
                Dim i As Integer = 0
                Do While (i <= num2)
                    If textArray2(i).Contains("lock_tran9_") Then
                        Trace.Write("Found file: ")
                        Trace.Write(textArray2(i))
                        Trace.Write(ChrW(13) & ChrW(10))
                        Dim strPath As String = Path.Combine(ImportMain.FileLocation, textArray2(i))
                        If File.Exists(strPath) Then
                            strPath = ImportMain.GetUniqueFileName(strPath)
                        End If
                        Trace.Write("Downloading to ")
                        Trace.Write(strPath)
                        Trace.Write(ChrW(13) & ChrW(10))
                        client.Get(strPath, textArray2(i))
                        Trace.WriteLine("Deleting file...")
                        client.Delete(textArray2(i))
                        list.Add(strPath)
                    Else
                        Trace.Write("Ignoring file: ")
                        Trace.Write(textArray2(i))
                        Trace.Write("... Bad naming convention..." & ChrW(13) & ChrW(10))
                    End If
                    i += 1
                Loop
                Trace.WriteLine("Logging off...")
                client.Quit()
                Trace.WriteLine("Process complete.")
                Return DirectCast(list.ToArray(GetType(String)), String())
            Catch exception1 As Exception
                ProjectData.SetProjectError(exception1)
                Dim exception As Exception = exception1
                Trace.WriteLine("Exception:")
                Trace.Write(exception.ToString)
                ProjectData.ClearProjectError()
            End Try
            Return Nothing
        End Function

        Private Shared Function GetUniqueFileName(ByVal originalFileName As String) As String
            Dim directoryName As String = Path.GetDirectoryName(originalFileName)
            Dim fileNameWithoutExtension As String = Path.GetFileNameWithoutExtension(originalFileName)
            Dim extension As String = Path.GetExtension(originalFileName)
            Dim i As Integer = 1
            Do While File.Exists(originalFileName)
                originalFileName = Path.Combine(directoryName, String.Concat(New Object() {fileNameWithoutExtension, " DUPLICATE (", i, ")", extension}))
                i += 1
            Loop
            Return originalFileName
        End Function

        Public Shared Function ImportFile(ByVal fileName As String, ByVal connectionString As String) As Integer
            Dim CurrentClientID As Integer
            Dim ClientIDs As New List(Of Integer)
            Dim num6 As Integer
            Dim num2 As Double = Double.Parse(PropertyHelper.Value("HoldCheckAmount"))
            Dim num3 As Double = Double.Parse(PropertyHelper.Value("HoldCheckDays"))
            Trace.WriteLine("--------------------------------")
            Trace.Write(DateTime.Now.ToString)
            Trace.Write(" - Importing" & ChrW(13) & ChrW(10))
            Trace.Write("File: ")
            Trace.Write(fileName)
            Trace.Write(ChrW(13) & ChrW(10) & "Connection String: ")
            Trace.Write(connectionString)
            Trace.Write(ChrW(13) & ChrW(10) & "--------------------------------" & ChrW(13) & ChrW(10))
            Dim buffer As Byte() = ImportMain.GenerateMD5(fileName)
            Using connection As SqlConnection = New SqlConnection(connectionString)
                Using cmd As SqlCommand = connection.CreateCommand
                    cmd.CommandTimeout = 300
                    cmd.CommandText = "SELECT ImportId FROM tblImport WHERE MD5=@MD5"
                    DatabaseHelper.AddParameter(cmd, "MD5", buffer)
                    connection.Open()
                    Dim list As New List(Of String)
                    Using reader As SqlDataReader = cmd.ExecuteReader
                        Do While reader.Read
                            list.Add(reader.GetInt32(0).ToString)
                        Loop
                    End Using
                    If (list.Count > 0) Then
                        Trace.WriteLine("Found duplicate files:")
                        Dim num16 As Integer = (list.Count - 1)
                        Dim i As Integer = 0
                        Do While (i <= num16)
                            Trace.WriteLine(("ImportId = " & list.Item(i)))
                            i += 1
                        Loop
                        Trace.WriteLine("Skipping Import!")
                        cmd.Parameters.Clear()
                        DatabaseHelper.AddParameter(cmd, "Imported", DateTime.Now)
                        DatabaseHelper.AddParameter(cmd, "ImportedBy", &H18)
                        DatabaseHelper.AddParameter(cmd, "Database", "TEXT FILE")
                        DatabaseHelper.AddParameter(cmd, "Table", fileName)
                        DatabaseHelper.AddParameter(cmd, "Description", ("THIS FILE MATCHES PREVIOUS IMPORT " & String.Join(", ", list.ToArray)))
                        DatabaseHelper.AddParameter(cmd, "MD5", buffer)
                        cmd.CommandText = "INSERT INTO tblImport (Imported, ImportedBy, [Database], [Table], [Description], MD5) VALUES (@Imported, @ImportedBy, @Database, @Table, @Description, @MD5); SELECT CAST(SCOPE_IDENTITY() AS int)"
                        Return Conversions.ToInteger(cmd.ExecuteScalar)
                    End If
                End Using
            End Using
            Try
                Using reader2 As StreamReader = New StreamReader(fileName)
                    Using connection2 As SqlConnection = New SqlConnection(connectionString)
                        Dim text2 As String = reader2.ReadLine
                        Dim time As New DateTime(Integer.Parse(text2.Substring(4)), Integer.Parse(text2.Substring(0, 2)), Integer.Parse(text2.Substring(2, 2)))
                        Using command2 As SqlCommand = connection2.CreateCommand
                            command2.CommandTimeout = 300
                            DatabaseHelper.AddParameter(command2, "Imported", DateTime.Now)
                            DatabaseHelper.AddParameter(command2, "ImportedBy", &H18)
                            DatabaseHelper.AddParameter(command2, "Database", "TEXT FILE")
                            DatabaseHelper.AddParameter(command2, "Table", fileName)
                            DatabaseHelper.AddParameter(command2, "Description", "Daily lockbox report from Colonial Bank")
                            DatabaseHelper.AddParameter(command2, "MD5", buffer)
                            command2.CommandText = "INSERT INTO tblImport (Imported, ImportedBy, [Database], [Table], [Description], MD5) VALUES (@Imported, @ImportedBy, @Database, @Table, @Description, @MD5); SELECT CAST(SCOPE_IDENTITY() AS int)"
                            connection2.Open()
                            num6 = Conversions.ToInteger(command2.ExecuteScalar)
                        End Using
                        Dim message As String = Nothing
                        Dim num7 As Integer = 2
                        Dim num8 As Integer = 0
                        Dim num9 As Double = 0
                        Dim num10 As Integer = 0
                        Dim num5 As Double = 0
                        message = reader2.ReadLine
                        Do While (Not message Is Nothing)
                            If (IIf(((Not message Is Nothing) AndAlso (message.Length > 0)), 1, 0) <> 0) Then
                                Dim textArray As String() = message.Split(New Char() {","c})
                                If (textArray.Length = 4) Then
                                    Dim text3 As String = textArray(0)
                                    Dim num11 As Integer = Integer.Parse(textArray(1))
                                    Dim num12 As Double = Double.Parse(textArray(2))
                                    Dim text4 As String = textArray(3)
                                    If (num11 = 9) Then
                                        num11 = 3
                                    End If
                                    If (IIf(((Operators.CompareString(textArray(1), "13", False) = 0) OrElse (Operators.CompareString(textArray(1), "9", False) = 0)), 1, 0) <> 0) Then
                                        num8 += 1
                                        num9 = (num9 + num12)
                                        Dim num13 As Integer = 0
                                        Dim num14 As Integer = DataHelper.Nz_int(DataHelper.FieldLookup("tblClient", "ClientID", ("AccountNumber = '" & text3.Replace("u", "0").Replace("U", "0").TrimStart(New Char() {"0"c}) & "'")))
                                        CurrentClientID = num14
                                        If (num14 <> 0) Then
                                            Dim command3 As IDbCommand = ConnectionFactory.Create.CreateCommand
                                            command3.CommandTimeout = 300
                                            Try
                                                DatabaseHelper.AddParameter(command3, "ClientID", num14)
                                                DatabaseHelper.AddParameter(command3, "TransactionDate", time)
                                                DatabaseHelper.AddParameter(command3, "CheckNumber", text4)
                                                DatabaseHelper.AddParameter(command3, "Amount", num12)
                                                DatabaseHelper.AddParameter(command3, "EntryTypeID", 3)
                                                DatabaseHelper.AddParameter(command3, "ImportID", num6)
                                                If (num12 >= num2) Then
                                                    DatabaseHelper.AddParameter(command3, "Hold", time.AddDays(num3))
                                                    DatabaseHelper.AddParameter(command3, "HoldBy", &H18)
                                                End If
                                                DatabaseHelper.BuildInsertCommandText((command3), "tblRegister")
                                                command3.Connection.Open()
                                                num13 = command3.ExecuteNonQuery
                                            Finally
                                                DatabaseHelper.EnsureConnectionClosed(command3.Connection)
                                            End Try
                                            ClientHelper.CleanupRegister(num14)
                                            ClientIDs.Add(num14)
                                            'RegisterHelper.Rebalance(num14)
                                        End If
                                        If (num13 > 0) Then
                                            num10 += 1
                                            num5 = (num5 + num12)
                                        Else
                                            Trace.Write("Line ")
                                            Trace.Write(num7.ToString)
                                            Trace.Write(", Account ")
                                            Trace.Write(text3.ToString)
                                            Trace.Write(", Amount ")
                                            Trace.Write(num12.ToString)
                                            Trace.Write(", Check ")
                                            Trace.Write(text4.ToString)
                                            Trace.Write(" insert error: Client not found." & ChrW(13) & ChrW(10))
                                        End If
                                    Else
                                        Trace.Write("Line ")
                                        Trace.Write(num7.ToString)
                                        Trace.Write(", Account ")
                                        Trace.Write(text3.ToString)
                                        Trace.Write(", Amount ")
                                        Trace.Write(num12.ToString)
                                        Trace.Write(", Check ")
                                        Trace.Write(text4.ToString)
                                        Trace.Write(" invalid Entry Type ")
                                        Trace.Write(textArray(1))
                                        Trace.Write("." & ChrW(13) & ChrW(10))
                                    End If
                                Else
                                    Trace.Write("Line ")
                                    Trace.Write(num7.ToString)
                                    Trace.Write(" could not properly parse """)
                                    Trace.Write(message)
                                    Trace.Write(""". Skipping." & ChrW(13) & ChrW(10))
                                End If
                                num7 += 1
                            End If
                            message = reader2.ReadLine
                        Loop
                        Trace.WriteLine("Import complete.")
                        Trace.Write("Total line items: ")
                        Trace.Write(num8.ToString)
                        Trace.Write(", Amount: ")
                        Trace.Write(num9.ToString)
                        Trace.Write(ChrW(13) & ChrW(10) & "Successful line items: ")
                        Trace.Write(num10.ToString)
                        Trace.Write(", Amount: ")
                        Trace.Write(num5.ToString)
                        Trace.Write(ChrW(13) & ChrW(10) & "Unsuccessful line items: ")
                        Trace.Write((num8 - num10).ToString)
                        Trace.Write(", Amount: ")
                        Trace.Write((num9 - num5).ToString)
                        Trace.Write(ChrW(13) & ChrW(10))
                    End Using
                End Using

                Trace.Write("Assigning Negotiators..." & ChrW(13) & ChrW(10))

                Using cmd As New SqlCommand("", New SqlConnection(connectionString))
                    Using cmd.Connection
                        cmd.CommandTimeout = 3600

                        cmd.Connection.Open()

                        For Each id As Integer In ClientIDs
                            CurrentClientID = id
                            cmd.CommandText = "exec stp_AssignNegotiator " + id.ToString()

                            cmd.ExecuteNonQuery()
                        Next
                    End Using
                End Using

                Trace.Write("Assigned Negotiators For All " & ClientIDs.Count & " Clients." & ChrW(13) & ChrW(10))

                Return num6
            Catch exception1 As FileNotFoundException
                ProjectData.SetProjectError(exception1)
                Dim exception As FileNotFoundException = exception1
                Trace.Write("Could not find file:" & ChrW(13) & ChrW(10))
                Trace.Write(exception.ToString)
                ProjectData.ClearProjectError()
            Catch exception5 As IOException
                ProjectData.SetProjectError(exception5)
                Dim exception2 As IOException = exception5
                Trace.Write("IO error:" & ChrW(13) & ChrW(10))
                Trace.Write(exception2.ToString)
                ProjectData.ClearProjectError()
            Catch exception6 As SqlException
                ProjectData.SetProjectError(exception6)
                Dim exception3 As SqlException = exception6
                Trace.Write("Died Trying To Process Client: " & CurrentClientID & ChrW(13) & ChrW(10))
                Trace.Write("SQL Error:" & ChrW(13) & ChrW(10))
                Trace.Write(exception3.ToString)
                ProjectData.ClearProjectError()
            Catch exception7 As Exception
                ProjectData.SetProjectError(exception7)
                Dim exception4 As Exception = exception7
                Trace.Write("Exception:" & ChrW(13) & ChrW(10))
                Trace.Write(exception4.ToString)
                ProjectData.ClearProjectError()
            End Try
            Return 0
        End Function

        <STAThread()> _
        Public Shared Sub Main(ByVal parms As String())
            Using writer As TextWriter = New StreamWriter(ImportMain.LogFile, True)
                Trace.Listeners.Add(New TextWriterTraceListener(writer))
                If (IIf(((parms.Length > 0) AndAlso (Operators.CompareString(parms(0).Trim.ToLower, "-batch", False) = 0)), 1, 0) <> 0) Then
                    Dim ftpLatest As String()
                    Dim writer2 As StringWriter = Nothing
                    Dim listener As TextWriterTraceListener = Nothing
                    Dim writer3 As StringWriter = Nothing
                    Dim listener2 As TextWriterTraceListener = Nothing
                    writer3 = New StringWriter
                    listener2 = New TextWriterTraceListener(writer3)
                    Trace.Listeners.Add(listener2)
                    If (IIf(((Not ImportMain.LogEmailTo Is Nothing) AndAlso (ImportMain.LogEmailTo.Length > 0)), 1, 0) <> 0) Then
                        writer2 = New StringWriter
                        listener = New TextWriterTraceListener(writer2)
                        Trace.Listeners.Add(listener)
                    End If
                    If (parms.Length > 1) Then
                        ftpLatest = parms(1).Replace("""", "").Split(New Char() {","c})
                    Else
                        ftpLatest = ImportMain.GetFtpLatest
                    End If
                    Dim list As New List(Of Integer)
                    If (Not ftpLatest Is Nothing) Then
                        Dim num4 As Integer = (ftpLatest.Length - 1)
                        Dim i As Integer = 0
                        Do While (i <= num4)
                            Dim item As Integer = ImportMain.ImportFile(ftpLatest(i), ImportMain.ConnectionString)
                            If (item <> 0) Then
                                list.Add(item)
                            End If
                            i += 1
                        Loop
                        Trace.WriteLine("Goodbye.")
                    End If
                    If (Not listener Is Nothing) Then
                        listener.Flush()
                        writer2.Flush()
                        Trace.Listeners.Remove(listener)
                        listener.Dispose()
                        ImportMain.EmailLog(writer2.ToString)
                        writer2.Close()
                    End If
                    If (Not listener2 Is Nothing) Then
                        listener2.Flush()
                        writer3.Flush()
                        Trace.Listeners.Remove(listener2)
                        listener2.Dispose()
                        Dim num3 As Integer
                        For Each num3 In list
                            ImportLogHelper.Insert(num3, writer3.ToString)
                        Next
                        writer3.Close()
                    End If
                Else
                    Application.Run(New ImportMainForm)
                End If
            End Using
        End Sub


        ' Properties
        Public Shared ReadOnly Property AfterImportStoredProc As String
            Get
                Return ConfigurationSettings.AppSettings.Item("afterImportStoredProc")
            End Get
        End Property

        Public Shared ReadOnly Property AfterImportTimeout As Integer
            Get
                Dim s As String = ConfigurationSettings.AppSettings.Item("afterImportTimeout")
                If (IIf(((Not s Is Nothing) AndAlso (s.Length > 0)), 1, 0) <> 0) Then
                    Try 
                        Return Integer.Parse(s)
                    Catch exception1 As Exception
                        ProjectData.SetProjectError(exception1)
                        Dim exception As Exception = exception1
                        ProjectData.ClearProjectError
                        Return &H708
                        ProjectData.ClearProjectError
                    End Try
                End If
                Return &H708
            End Get
        End Property

        Public Shared ReadOnly Property ConnectionString As String
            Get
                Return ConfigurationSettings.AppSettings.Item("connectionString")
            End Get
        End Property

        Public Shared ReadOnly Property FileLocation As String
            Get
                Return ConfigurationSettings.AppSettings.Item("fileLocation")
            End Get
        End Property

        Public Shared ReadOnly Property FtpFolder As String
            Get
                Return ConfigurationSettings.AppSettings.Item("ftpFolder")
            End Get
        End Property

        Public Shared ReadOnly Property FtpPassword As String
            Get
                Return ConfigurationSettings.AppSettings.Item("ftpPassword")
            End Get
        End Property

        Public Shared ReadOnly Property FtpServer As String
            Get
                Return ConfigurationSettings.AppSettings.Item("ftpServer")
            End Get
        End Property

        Public Shared ReadOnly Property FtpUserName As String
            Get
                Return ConfigurationSettings.AppSettings.Item("ftpUserName")
            End Get
        End Property

        Public Shared ReadOnly Property LogEmailTo As String
            Get
                Return ConfigurationSettings.AppSettings.Item("logEmailTo")
            End Get
        End Property

        Public Shared ReadOnly Property LogFile As String
            Get
                Dim text2 As String = ConfigurationSettings.AppSettings.Item("logFile")
                If (IIf(((Not text2 Is Nothing) AndAlso (text2.Length <> 0)), 1, 0) <> 0) Then
                    Return text2
                End If
                Return Path.Combine(Path.GetDirectoryName(Assembly.GetEntryAssembly.Location), "Import.log")
            End Get
        End Property

        Public Shared ReadOnly Property SmtpServer As String
            Get
                Dim text2 As String = ConfigurationSettings.AppSettings.Item("smtpServer")
                If (IIf(((Not text2 Is Nothing) AndAlso (text2.Length <> 0)), 1, 0) <> 0) Then
                    Return text2
                End If
                Return "localhost"
            End Get
        End Property

    End Class
End Namespace

File: Slf.Dms.Service.ICBImport/ImportMain.cs
using System;
using System.Configuration;
using System.Collections;
using System.Data;
using System.Data.SqlClient;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using System.Windows.Forms;
using System.Web.Mail;
using System.Collections.Generic;

using Drg.Util.DataAccess;
using Drg.Util.DataHelpers;

using EnterpriseDT.Net.Ftp;

// TODO: create import header table for unique ids
// TODO: assign unique id to each import to enable rollback
// TODO: write to the Exception table

namespace Slf.Dms.Service.ICBImport
{
	public class ImportMain
	{
		[STAThread]
		static void Main(string[] parms) 
		{
			using (TextWriter tw = new StreamWriter(LogFile, true))
			{
				// Hook up a log file trace writer
				Trace.Listeners.Add(new TextWriterTraceListener(tw));

				if (parms.Length > 0 && parms[0].ToLower() == "-batch")
				{
					StringWriter emailWriter = null;
					TextWriterTraceListener emailListener = null;

                    StringWriter ImportWriter = null;
                    TextWriterTraceListener ImportListener = null;

                    ImportWriter = new StringWriter();
                    ImportListener = new TextWriterTraceListener(ImportWriter);

                    Trace.Listeners.Add(ImportListener);

                    if (LogEmailTo != null && LogEmailTo.Length > 0)
					{
						// Hook up a listener to compile the log for emailing
						emailWriter = new StringWriter();
						emailListener = new TextWriterTraceListener(emailWriter);

						Trace.Listeners.Add(emailListener);
					}

					string[] fileNames = GetFtpLatest();

                    List<int> ImportIds = new List<int>();

					if (fileNames != null)
					{
                        for (int i = 0; i < fileNames.Length; i++)
                        {
                            int ImportID = ImportFile(fileNames[i], ConnectionString);

                            if (ImportID != 0)
                                ImportIds.Add(ImportID);
                        }

						if (fileNames.Length > 0)
						{
							ExecuteAfterImportStoredProc(ConnectionString);
						}
					}

					if (emailListener != null)
					{
						emailListener.Flush();
						emailWriter.Flush();

						Trace.Listeners.Remove(emailListener);
						emailListener.Dispose();

						EmailLog(emailWriter.ToString());

						emailWriter.Close();
					}

                    if (ImportListener != null)
                    {
                        ImportListener.Flush();
                        ImportWriter.Flush();

                        Trace.Listeners.Remove(ImportListener);
                        ImportListener.Dispose();

                        // insert entire log against each import
                        foreach (int ImportID in ImportIds)
                        {
                            ImportLogHelper.Insert(ImportID, ImportWriter.ToString());
                        }

                        ImportWriter.Close();
                    }
				}
				else
				{
					Application.Run(new ImportMainForm());
				}
			}
		}

		static string[] GetFtpLatest()
		{
			Trace.WriteLine("--------------------------------");
			Trace.Write(DateTime.Now.ToString());
			Trace.Write(" - Retrieving from FTP\r\n");
			Trace.Write("Server: ");
			Trace.Write(FtpServer);
			Trace.Write("\r\nFolder: ");
			Trace.Write(FtpFolder);
			Trace.Write("\r\n--------------------------------\r\n");

			try 
			{
				Trace.WriteLine("Connecting...");
				FTPClient ftp = new FTPClient(FtpServer);

				Trace.WriteLine("Logging in...");
				ftp.Login(FtpUserName, FtpPassword);

				Trace.WriteLine("Setting up passive, ASCII transfers...");
                ftp.ConnectMode = FTPConnectMode.PASV;
				ftp.TransferType = FTPTransferType.ASCII;

				if (FtpFolder != null && FtpFolder.Length > 0)
					ftp.ChDir(FtpFolder);

				string[] files = ftp.Dir();

				ArrayList importFiles = new ArrayList(files.Length);

				for (int i = 0; i < files.Length; i++)
				{
					if (files[i].StartsWith("seidlock_tran9_"))
					{
						Trace.Write("Found file: ");
						Trace.Write(files[i]);
						Trace.Write("\r\n");

						string localPath = Path.Combine(FileLocation, files[i]);

						Trace.Write("Downloading to ");
						Trace.Write(localPath);
						Trace.Write("\r\n");

						ftp.Get(localPath, files[i]);

						Trace.WriteLine("Deleting file...");
						ftp.Delete(files[i]);

						importFiles.Add(localPath);
					}
					else
					{
						Trace.Write("Ignoring file: ");
						Trace.Write(files[i]);
						Trace.Write("... Bad naming convention...\r\n");
					}
				}

				// Shut down client                
				Trace.WriteLine("Logging off...");
				ftp.Quit();
            
				Trace.WriteLine("Process complete.");

				return (string[])importFiles.ToArray(typeof(string));
			} 
			catch (Exception e) 
			{
				Trace.WriteLine("Exception:");
				Trace.Write(e.ToString());
			}

			return null;
		}

		public static int ImportFile(string fileName, string connectionString)
		{
            double HoldCheckAmount = double.Parse(PropertyHelper.Value("HoldCheckAmount"));
            double HoldCheckDays = double.Parse(PropertyHelper.Value("HoldCheckDays"));

			Trace.WriteLine("--------------------------------");
			Trace.Write(DateTime.Now.ToString());
			Trace.Write(" - Importing\r\n");
			Trace.Write("File: ");
			Trace.Write(fileName);
			Trace.Write("\r\nConnection String: ");
			Trace.Write(connectionString);
			Trace.Write("\r\n--------------------------------\r\n");

			try
			{
                int ImportID;
                
                using (StreamReader s = new StreamReader(fileName))
				using (SqlConnection cn = new SqlConnection(connectionString))
				{
					//string dateString = s.ReadLine();
                    string file = Path.GetFileName(fileName);

                    DateTime date = new DateTime(int.Parse(file.Substring(15, 4)),
                        int.Parse(file.Substring(9, 2)),
                        int.Parse(file.Substring(12, 2)));

					using (SqlCommand ImportCommand = cn.CreateCommand())
					{
                        DatabaseHelper.AddParameter(ImportCommand, "Imported", DateTime.Now);
                        DatabaseHelper.AddParameter(ImportCommand, "ImportedBy", 24); // user 24 is Import Engine
                        DatabaseHelper.AddParameter(ImportCommand, "Database", "TEXT FILE");
                        DatabaseHelper.AddParameter(ImportCommand, "Table", fileName);
                        DatabaseHelper.AddParameter(ImportCommand, "Description", "Daily lockbox report from Inland Community Bank");

                        ImportCommand.CommandText = "INSERT INTO tblImport (Imported, ImportedBy, [Database], " 
                            + "[Table], [Description]) VALUES (@Imported, @ImportedBy, @Database, @Table, " 
                            + "@Description); SELECT CAST(SCOPE_IDENTITY() AS int)";

                        cn.Open();

                        ImportID = (int)ImportCommand.ExecuteScalar();
					}

					string Line = null;
					int LineNumber = 2;
					int TotalLines = 0;
					double TotalAmount = 0;
					int TotalSuccessfulLines = 0;
					double TotalSuccessfulAmount = 0;

					while ((Line = s.ReadLine()) != null)
					{
						if (Line.Length > 45)
						{
                            string accountNumber = Line.Substring(7, 9);            // 8-16 Account No
                            int transCode = int.Parse(Line.Substring(65, 3));       // 66-68 Trans Code
                            double amount = double.Parse(Line.Substring(68, 10));   // 69-78 Amount
                            int checkNumber = int.Parse(Line.Substring(78, 8));     // 79-86 CheckNo
                            string postDate = Line.Substring(88, 8);                // 89-96 Posted Date
                            DateTime transDate = DateTime.Parse(postDate.Substring(0, 2) + "/" + postDate.Substring(2, 2) + "/" + postDate.Substring(postDate.Length - 4, 4));

                            int entryTypeId;

                            switch (transCode)
                            {
                                case 8:
                                case 9:
                                case 13:
                                default:
                                    // deposit
                                    entryTypeId = 3;

                                    break;
                                case 43:
                                    // program fees payment
                                    entryTypeId = 1;

                                    break;
                                case 52:
                                    // settlement/client withdrawl
                                    entryTypeId = 18;

                                    break;
                            }

							TotalLines++;
							TotalAmount += amount;

							int RecordsUpdated = 0;
							
                            // lookup client
                            int ClientID = DataHelper.Nz_int(DataHelper.FieldLookup("tblClient", 
                                "ClientID", "AccountNumber = '" + accountNumber + "'"));

                            // insert transaction against client
                            if (ClientID != 0)
                            {
                                // create command
                                IDbCommand cmd = ConnectionFactory.Create().CreateCommand();

                                try
                                {
                                    DatabaseHelper.AddParameter(cmd, "ClientID", ClientID);
                                    DatabaseHelper.AddParameter(cmd, "TransactionDate", transDate);
                                    DatabaseHelper.AddParameter(cmd, "CheckNumber", checkNumber);
                                    DatabaseHelper.AddParameter(cmd, "Amount", amount);
                                    DatabaseHelper.AddParameter(cmd, "EntryTypeID", entryTypeId);
                                    DatabaseHelper.AddParameter(cmd, "ImportID", ImportID);

                                    if (amount >= HoldCheckAmount)
                                    {
                                        DatabaseHelper.AddParameter(cmd, "Hold", date.AddDays(HoldCheckDays));
                                    }

                                    DatabaseHelper.BuildInsertCommandText(ref cmd, "tblRegister");

                                    cmd.Connection.Open();

                                    RecordsUpdated = cmd.ExecuteNonQuery();
                                }
                                finally
                                {
                                    DatabaseHelper.EnsureConnectionClosed(cmd.Connection);
                                }

                                // run payment manager on client
                                ClientHelper.CleanupRegister(ClientID);
                            }

							if (RecordsUpdated > 0)
							{
								TotalSuccessfulLines++;
								TotalSuccessfulAmount += amount;

								Trace.Write("Line ");
								Trace.Write(LineNumber.ToString());
								Trace.Write(", Account ");
								Trace.Write(accountNumber.ToString());
								Trace.Write(", Amount ");
								Trace.Write(amount.ToString());
								Trace.Write(", Check ");
								Trace.Write(checkNumber.ToString());
								Trace.Write(" successfully inserted ");
								Trace.Write(RecordsUpdated.ToString());
								Trace.Write(" record.\r\n");
							}
							else
							{
								Trace.Write("Line ");
								Trace.Write(LineNumber.ToString());
								Trace.Write(", Account ");
								Trace.Write(accountNumber.ToString());
								Trace.Write(", Amount ");
								Trace.Write(amount.ToString());
								Trace.Write(", Check ");
								Trace.Write(checkNumber.ToString());
								Trace.Write(" insert error: Client_Id not found.\r\n");

								// TODO: write exception log table
								//InsertException(
							}
                        }
						else
						{
							Trace.Write("Line ");
							Trace.Write(LineNumber.ToString());
							Trace.Write(" could not properly parse \"");
							Trace.Write(Line);
							Trace.Write("\". Skipping.\r\n");
						}

						LineNumber++;
					}

					Trace.WriteLine("Import complete.");
					Trace.Write("Total line items: ");
					Trace.Write(TotalLines.ToString());
					Trace.Write(", Amount: ");
					Trace.Write(TotalAmount.ToString());
					Trace.Write("\r\nSuccessful line items: ");
					Trace.Write(TotalSuccessfulLines.ToString());
					Trace.Write(", Amount: ");
					Trace.Write(TotalSuccessfulAmount.ToString());
					Trace.Write("\r\nUnsuccessful line items: ");
					Trace.Write((TotalLines - TotalSuccessfulLines).ToString());
					Trace.Write(", Amount: ");
					Trace.Write((TotalAmount - TotalSuccessfulAmount).ToString());
					Trace.Write("\r\n");
				}

                return ImportID;
			}
			catch (FileNotFoundException ex)
			{
				Trace.Write("Could not find file:\r\n");
				Trace.Write(ex.ToString());
			}
			catch (IOException ex)
			{
				Trace.Write("IO error:\r\n");
				Trace.Write(ex.ToString());
			}
			catch (SqlException ex)
			{
				Trace.Write("SQL Error:\r\n");
				Trace.Write(ex.ToString());
			}
			catch (Exception ex)
			{
				Trace.Write("Exception:\r\n");
				Trace.Write(ex.ToString());
			}

            return 0;
		}

		public static void ExecuteAfterImportStoredProc(string connectionString)
		{
			Trace.WriteLine("--------------------------------");
			Trace.Write(DateTime.Now.ToString());
			Trace.Write(" - Updating transactions\r\n");
			Trace.Write("Stored Proc: ");
			Trace.Write(AfterImportStoredProc);
			Trace.Write("\r\nTimeout: ");
			Trace.Write(AfterImportTimeout);
			Trace.Write("\r\n--------------------------------\r\n");

			if (AfterImportStoredProc != null && AfterImportStoredProc.Length > 0)
			{
				using (SqlConnection cn = new SqlConnection(connectionString))
				using (SqlCommand cmd = cn.CreateCommand())
				{
					cmd.CommandText = AfterImportStoredProc;
					cmd.CommandType = CommandType.StoredProcedure;

					cmd.CommandTimeout = AfterImportTimeout;

					cn.Open();

					Trace.WriteLine("Executing...");
					cmd.ExecuteNonQuery();
					Trace.WriteLine("Completed");
				}
			}
			else
			{
				Trace.WriteLine("No stored procedure specified. Aborting.");
			}
		}

		static void EmailLog(string log)
		{
			if (LogEmailTo != null && LogEmailTo.Length > 0)
			{
				MailMessage msg = new MailMessage();

				msg.From = "service@dmsisupport.com";
				msg.To = LogEmailTo;
				msg.Subject = "Import Log for " + DateTime.Now.ToShortDateString();
				msg.Body = log;

				SmtpMail.SmtpServer = SmtpServer;

				try
				{
					SmtpMail.Send(msg);
				}
				catch (Exception ex)
				{
					Trace.WriteLine("Exception sending mail:");
					Trace.WriteLine(ex.ToString());
				}
			}
		}

		public static string ConnectionString
		{
			get
			{
				return ConfigurationSettings.AppSettings["connectionString"];
			}
		}

		public static string FtpServer
		{
			get
			{
				return ConfigurationSettings.AppSettings["ftpServer"];
			}
		}

		public static string FtpFolder
		{
			get
			{
				return ConfigurationSettings.AppSettings["ftpFolder"];
			}
		}

		public static string FtpUserName
		{
			get
			{
				return ConfigurationSettings.AppSettings["ftpUserName"];
			}
		}

		public static string FtpPassword
		{
			get
			{
				return ConfigurationSettings.AppSettings["ftpPassword"];
			}
		}

		public static string LogFile
		{
			get
			{
				string logFile = ConfigurationSettings.AppSettings["logFile"];

				if (logFile == null || logFile.Length == 0)
					logFile = Path.Combine(Path.GetDirectoryName(Assembly.GetEntryAssembly().Location), "Import.log");

				return logFile;
			}
		}
		
		public static string LogEmailTo
		{
			get
			{
				return ConfigurationSettings.AppSettings["logEmailTo"];
			}
		}

		public static string SmtpServer
		{
			get
			{
				string server = ConfigurationSettings.AppSettings["smtpServer"];

				if (server == null || server.Length == 0)
					return "localhost";
				else
					return server;
			}
		}

		public static string FileLocation
		{
			get
			{
				return ConfigurationSettings.AppSettings["fileLocation"];
			}
		}

		public static string AfterImportStoredProc
		{
			get
			{
				return ConfigurationSettings.AppSettings["afterImportStoredProc"];
			}
		}

		public static int AfterImportTimeout
		{
			get
			{
				string timeout = ConfigurationSettings.AppSettings["afterImportTimeout"];

				if (timeout != null && timeout.Length > 0)
				{
					try
					{
						return int.Parse(timeout);
					}
					catch
					{
						return 1800;
					}
				}
				else
				{
					return 1800;
				}
			}
		}
	}
}